hpx/parallel/algorithms/for_loop.hpp

```diff
@@ -206,19 +206,19 @@ namespace hpx { namespace parallel { inline namespace v2
                 if (size == 0)
                     return util::detail::algorithm_result<ExPolicy>::get();
 
-                auto && args =
-                    hpx::util::forward_as_tuple(std::forward<Ts>(ts)...);
-
                 // we need to decay copy here to properly transport everything
                 // to a GPU device
                 typedef
                     hpx::util::tuple<typename hpx::util::decay<Ts>::type...>
                     args_type;
 
+                args_type args =
+                    hpx::util::forward_as_tuple(std::forward<Ts>(ts)...);
+
                 return util::partitioner<ExPolicy>::call_with_index(
                     policy, first, size, stride,
                     part_iterations<F, S, args_type>{
-                        std::forward<F>(f), stride, std::move(args)
+                        std::forward<F>(f), stride, args
                     },
```

Comment:
AFAICS, this code causes the data to be copied twice. Once when creating `args` and once while calling `call_with_index`. Is there any way to avoid that?

Reply:
Yes, that's a potential problem indeed. I don't know how to avoid that. `args` is needed twice, once for `part_iterations` and then in the finalizer.

Reply:
Couldn't `args_type` be a tuple of references instead?

Reply:
No, they need to decay-copied at least once to stay in scope long enough.

Reply:
Ok, so let's copy them once and not thrice.

Reply:
I wouldn't know how...
The first one, is forwarding the arguments into a local tuple, which then needs to be passed to the iteration function and the exit function, which both need a copy of the arguments.

Reply:
Well, the exit function is guaranteed to outlive everything else, right? So why not hold the copy there and let the iteration refer to those copies? Would that work?

Reply:
Hmm, this would require some significant refactoring of the partitioner. I would prefer doing that in another PR though, but I guess it would work in principle

Reply:
Ok, the issue of having multiple copies should be fixed with the latest commit.

Reply:
Perfect, thanks!